### PR TITLE
build: Use specified branch of arrow-rs with workaround to invalid offset buffers from Java Arrow

### LINKED ISSUE #236: Use specified branch of arrow-rs with workaround to invalid offset buffers from Java Arrow
### What is the problem the feature request solves?

To unblock #95, as per discussion with @andygrove, I will go to use a branch of  my forked arrow-rs repo to add a workaround at Rust code that accepts the invalid offset buffers passed from Java Arrow. It is hacky and the buffer is invalid, but without that we cannot easily move forward because the Java Arrow issue.

### Describe the potential solution

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by viirya**
## Which issue does this PR close?

Closes #236.

### CODE REVIEW

**Comment by viirya** [comment]
With the specified arrow-rs and DataFusion forked repos which include a hacky workaround, I don't see 

```
General execution error with reason org.apache.comet.CometNativeException: Fail to process Arrow array with reason C Data interface error: The external buffer at position 1 is null...
```

the error showing anymore from TPCDSQuerySuite. But there are a few query failures unrelated to that.

File: core/Cargo.toml [hunk 1]
```diff
@@ -29,12 +29,12 @@ include = [
...
-arrow-data = { version = "~50.0.0" }
-arrow-schema = { version = "~50.0.0" }
-arrow-string = { version = "~50.0.0" }
-parquet = { version = "~50.0.0", default-features = false, features = ["experimental"] }
+arrow = { git = "https://github.com/viirya/arrow-rs.git", rev = "3f1ae0c", features = ["prettyprint", "ffi", "chrono-tz"] }
+arrow-array = { git = "https://github.com/viirya/arrow-rs.git", rev = "3f1ae0c" }
+arrow-data = { git = "https://github.com/viirya/arrow-rs.git", rev = "3f1ae0c" }
+arrow-schema = { git = "https://github.com/viirya/arrow-rs.git", rev = "3f1ae0c" }
+arrow-string = { git = "https://github.com/viirya/arrow-rs.git", rev = "3f1ae0c" }
+parquet = { git = "https://github.com/viirya/arrow-rs.git", rev = "3f1ae0c", default-features = false, features = ["experimental"] }
```
Comment (viirya) [inline core/Cargo.toml:37]: This patch basically switches to use the specified branch in my forked repo. It adds a hacky fix to the Java Arrow bug. Once the Java Arrow bug fix is merged and released, we can restore this back.

File: core/Cargo.toml [hunk 2]
```diff
@@ -66,9 +66,10 @@ itertools = "0.11.0"
 chrono = { version = "0.4", default-features = false, features = ["clock"] }
 chrono-tz = { version = "0.8" }
 paste = "1.0.14"
-datafusion-common = { version = "36.0.0" }
-datafusion = { default-features = false, version = "36.0.0", features = ["unicode_expressions"] }
-datafusion-physical-expr = { version = "36.0.0", default-features = false , features = ["unicode_expressions"] }
+datafusion-common = { git = "https://github.com/viirya/arrow-datafusion.git", rev = "111a940" }
+datafusion = { default-features = false, git = "https://github.com/viirya/arrow-datafusion.git", rev = "111a940", features = ["unicode_expressions"] }
+datafusion-functions = { git = "https://github.com/viirya/arrow-datafusion.git", rev = "111a940" }
+datafusion-physical-expr = { git = "https://github.com/viirya/arrow-datafusion.git", rev = "111a940", default-features = false, features = ["unicode_expressions"] }
```
Comment (viirya) [inline core/Cargo.toml:72]: We also need to use the specified version of arrow-rs in DataFusion, otherwise there will be conflicts.

**Review by viirya — COMMENTED** [review]
Except for arrow-rs and DataFusion crates changes, other changes are for API changes in DataFusion, including new built-in scalar resolution, new ExecutionPlan API, etc.

**Comment by viirya** [comment]
cc @sunchao @andygrove

**Comment by viirya** [comment]
> I tried to follow all the discussions of the related issues/PRs, and want to make sure I understand the issues/situations correctly. Per my understanding, please correct me if I'm wrong:
> 
> 1. Java Arrow's C data interface is exporting empty(/null) arraybuf for empty offset buffer, which will be fixed in [GH-40038: [Java] Export non empty offset buffer for variable-size layout through C Data Interface arrow#40043](https://github.com/apache/arrow/pull/40043).

Correct.

> 2. arrow-rs's implementation would allow empty offset buffer per discussion and its follow-up PRs: [Ensure there is a single zero in the offsets buffer for an empty ListArray. arrow-rs#1620](https://github.com/apache/arrow-rs/issues/1620) and https://github.com/apache/arrow-rs/pull/2836/files
>    
>    * But the arrow-rs ffi interface doesn't allow empty offset yet? So we have to patch arrow-rs in https://github.com/viirya/arrow-rs/commits/workaround_java_offset_bufs/ ?
>    * Will the upstream arrow-rs to pick that change/fix?
> 
> If the upstream arrow-rs is going to pick-up the fix, is it possible for us to wait a bit and use arrow-rs upstream directly? I'm a bit of worried about the stability/quality of using the main branch of arrow-rs directly.

No. As empty offset buffer is invalid, we are not going to merge the fix into arrow-rs. It is a hacky fix temporarily only before we have the fix at Java Arrow.

**Comment by viirya** [comment]
> Also, do you think it's a viable option to special handle offset buf in Comet's Java side @viirya ?
> 
> I just did a quick browsing of Arrow's export code, I think it should be doable to define a special ArrayExporter of Java Arrow c's exporter and write single value buffer for empty offset buffer in that exporter?

Looks like it is feasible to have a custom `ArrayExporter`. But it is also a hacky fix. We need to manually handle these offset-layouted arrays.

**Comment by viirya** [comment]
@sunchao @andygrove @parthchandra

**Review by andygrove — APPROVED** [review]
LGTM. Do we have an issue to track switching back to a versioned release of DataFusion/Arrow once Arrow Java 16 is released?

**Pull request merged**